File: plugins/tools/r_umap/runner.R
#!/usr/bin/env Rscript
# R UMAP dimensionality reduction workspace tool

args <- commandArgs(trailingOnly = TRUE)

tool_kind <- "umap"
input_path <- NULL
output_path <- NULL
check_deps <- FALSE
n_components <- 2L
n_neighbors <- 15L

i <- 1
while (i <= length(args)) {
  if (args[i] == "--tool-kind" && i + 1 <= length(args)) {
    tool_kind <- args[i + 1]
    i <- i + 2
  } else if (args[i] == "--input" && i + 1 <= length(args)) {
    input_path <- args[i + 1]
    i <- i + 2
  } else if (args[i] == "--output" && i + 1 <= length(args)) {
    output_path <- args[i + 1]
    i <- i + 2
  } else if (args[i] == "--check-deps") {
    check_deps <- TRUE
    i <- i + 1
  } else if (args[i] == "--n-components" && i + 1 <= length(args)) {
    n_components <- as.integer(args[i + 1])
    i <- i + 2
  } else if (args[i] == "--n-neighbors" && i + 1 <= length(args)) {
    n_neighbors <- as.integer(args[i + 1])
    i <- i + 2
  } else {
    i <- i + 1
  }
}

required <- c("ggplot2", "readr", "umap")
missing <- required[!vapply(required, requireNamespace, quietly = TRUE, FUN.VALUE = logical(1))]

if (check_deps) {
  if (length(missing) > 0) {
    cat(jsonlite::toJSON(list(status = "missing", packages = missing), auto_unbox = TRUE), "\n")
    quit(status = 2)
  }
  cat(jsonlite::toJSON(list(status = "ok", packages = required), auto_unbox = TRUE), "\n")
  quit(status = 0)
}

if (length(missing) > 0) {
  cat("Missing optional R dependencies:", paste(missing, collapse = ", "), "\n")
  cat("Install them in your workspace R library before running this tool.\n")
  quit(status = 2)
}

suppressPackageStartupMessages({
  library(ggplot2)
  library(readr)
  library(umap)
})

if (!is.null(input_path) && file.exists(input_path)) {
  df <- read_delim(input_path, delim = ifelse(grepl("\\.tsv$", input_path), "\t", ","), show_col_types = FALSE)
} else {
  set.seed(42)
  df <- as.data.frame(matrix(rnorm(100), nrow = 20, ncol = 5))
  colnames(df) <- paste0("feature_", seq_len(ncol(df)))
  rownames(df) <- paste0("sample_", seq_len(nrow(df)))
}

numeric_cols <- vapply(df, is.numeric, logical(1))
mat <- as.matrix(df[, numeric_cols, drop = FALSE])

umap_config <- umap.defaults
umap_config$n_components <- n_components
umap_config$n_neighbors <- n_neighbors

umap_result <- umap(mat, config = umap_config)
embedding <- umap_result$layout
colnames(embedding) <- paste0("UMAP", seq_len(ncol(embedding)))
rownames(embedding) <- rownames(mat)

if (!is.null(output_path)) {
  if (grepl("\\.(png|jpg|svg)$", output_path)) {
    png(output_path, width = 800, height = 600, res = 150)
    plot_df <- as.data.frame(embedding)
    p <- ggplot(plot_df, aes(x = UMAP1, y = UMAP2)) +
      geom_point(size = 3, alpha = 0.7) +
      theme_minimal() +
      ggtitle("UMAP Projection")
    print(p)
    dev.off()
  } else {
    writeLines(jsonlite::toJSON(as.data.frame(embedding), pretty = TRUE), output_path)
  }
  cat(jsonlite::toJSON(list(status = "ok", output = output_path, shape = dim(embedding)), auto_unbox = TRUE), "\n")
} else {
  cat(jsonlite::toJSON(list(
    status = "ok",
    shape = dim(embedding),
    message = "UMAP embedding computed. Use --output to save."
  ), auto_unbox = TRUE), "\n")
}

File: plugins/tools/umap/runner.py
#!/usr/bin/env python3
"""Python UMAP dimensionality reduction workspace tool."""

from __future__ import annotations

import argparse
import importlib.util
import json
import sys
from pathlib import Path


def missing(packages: list[str]) -> list[str]:
    return [pkg for pkg in packages if importlib.util.find_spec(pkg) is None]


def main() -> int:
    parser = argparse.ArgumentParser(description="SuperMedicine Python UMAP tool")
    parser.add_argument("--input", default=None, help="CSV/TSV data matrix path")
    parser.add_argument("--output", default=None, help="Output JSON or PNG path")
    parser.add_argument("--tool-kind", default="umap")
    parser.add_argument("--check-deps", action="store_true")
    parser.add_argument("--n-components", type=int, default=2, help="Number of UMAP dimensions")
    parser.add_argument("--n-neighbors", type=int, default=15, help="Number of neighbors for UMAP")
    args = parser.parse_args()

    required = ["pandas", "matplotlib", "umap"]
    if args.check_deps:
        unavailable = missing(required)
        if unavailable:
            print(json.dumps({"status": "missing", "packages": unavailable}))
            return 2
        print(json.dumps({"status": "ok", "packages": required}))
        return 0

    unavailable = missing(required)
    if unavailable:
        print(
            f"Missing optional Python dependencies: {', '.join(unavailable)}",
            file=sys.stderr,
        )
        print("Install them before running this tool.", file=sys.stderr)
        return 2

    try:
        import matplotlib
        matplotlib.use("Agg")
        import matplotlib.pyplot as plt
        import numpy as np
        import pandas as pd
        import umap  # type: ignore[import-not-found]

        if args.input:
            input_path = Path(args.input)
            if not input_path.is_file():
                raise FileNotFoundError(f"input file not found: {args.input}")
            sep = "\t" if input_path.suffix.lower() in (".tsv", ".tab") else ","
            df = pd.read_csv(input_path, sep=sep)
        else:
            np.random.seed(42)
            df = pd.DataFrame(
                np.random.randn(20, 5),
                columns=[f"feature_{i}" for i in range(5)],
                index=[f"sample_{i}" for i in range(20)],
            )

        numeric_df = df.select_dtypes(include="number")
        if numeric_df.empty:
            raise ValueError("no numeric columns found for UMAP")

        reducer = umap.UMAP(n_components=args.n_components, n_neighbors=args.n_neighbors)
        embedding = reducer.fit_transform(numeric_df.values)

        embedding_df = pd.DataFrame(
            embedding,
            columns=[f"UMAP{i+1}" for i in range(args.n_components)],
            index=numeric_df.index,
        )

        if args.output:
            output_path = Path(args.output)
            output_path.parent.mkdir(parents=True, exist_ok=True)
            if output_path.suffix.lower() in (".png", ".jpg", ".svg"):
                fig, ax = plt.subplots(figsize=(8, 6))
                ax.scatter(embedding[:, 0], embedding[:, 1], s=50, alpha=0.7)
                for i, label in enumerate(embedding_df.index):
                    ax.annotate(label, (embedding[i, 0], embedding[i, 1]), fontsize=8)
                ax.set_xlabel("UMAP1")
                ax.set_ylabel("UMAP2")
                ax.set_title("UMAP Projection")
                fig.savefig(output_path, dpi=150, bbox_inches="tight")
                plt.close(fig)
            else:
                output_path.write_text(
                    embedding_df.to_json(orient="index", indent=2) + "\n",
                    encoding="utf-8",
                )
            result = {"status": "ok", "output": str(output_path), "shape": list(embedding.shape)}
        else:
            result = {
                "status": "ok",
                "shape": list(embedding.shape),
                "embedding_sample": embedding_df.head(5).to_dict(orient="index"),
                "message": "UMAP embedding computed. Use --output to save.",
            }
        print(json.dumps(result, ensure_ascii=False, indent=2))
        return 0
    except Exception as exc:
        print(json.dumps({"status": "error", "error": str(exc)}), file=sys.stderr)
        return 2


if __name__ == "__main__":
    raise SystemExit(main())
